firedrake/formmanipulation.py

```diff
@@ -52,15 +65,22 @@ def split(self, form, argument_indices):
         """
         args = form.arguments()
         self._arg_cache = {}
-        self.blocks = dict(enumerate(argument_indices))
+        self.blocks = dict(enumerate(map(as_tuple, argument_indices)))
         if len(args) == 0:
             # Functional can't be split
             return form
         if all(len(a.function_space()) == 1 for a in args):
             assert (len(idx) == 1 for idx in self.blocks.values())
             assert (idx[0] == 0 for idx in self.blocks.values())
             return form
+        # TODO find a way to distinguish empty Forms avoiding expand_derivatives
```

Comment:
I am not sure if there is a robust solution for this. `expand_derivatives()` might be expensive for large forms. Can we not handle zero form after calling tsfc_compile_form()? Do we do this here for Slate?

Reply:
That's why I put the comment there. But both `action` and `adjoint` call `expand_derivatives` by default. Maybe this is a new behaivour that was intruduced alongside BaseForm. If `expand_derivatives` were not called, Fieldsplit/SLATE/EquationBC would not reach the empty Forms.

Reply:
https://github.com/firedrakeproject/ufl/blob/bfa596dd3f0e121168305451badf6b24c98dd7f6/ufl/formoperators.py#L174

Reply:
I see. I think I am fine with this.